File: Objc/TapPayTokenPushExample/TapPayTokenPushExample/TokenPushViewController.m
//
//  TokenPushViewController.m
//  TapPayTokenPushExample
//
//  Created by Cherri_TapPay_LukeCho on 2022/8/9.
//

#import "TokenPushViewController.h"
#import "TermsViewController.h"
#import "GlobalFunction.h"

#define DEFAULT_TRAILING 59.0
#define PUSH_TOKENIZE_PATH @"https://sandbox.tappaysdk.com/tpc/tsp/token/push-tokenize"
#define PARTNER_KEY @"axgLG8z7vVa8mgIzdU1233rjBFAEmTti19NZd1pI"

typedef NS_ENUM(NSInteger, BindStatus) {
    BindStatusSuccess,
    BindStatusFail,
    BindStatusCancel
};

@interface TokenPushViewController ()<TermsViewControllerDelegate>

@property (weak, nonatomic) IBOutlet UIImageView *statusImageView;
@property (weak, nonatomic) IBOutlet UILabel *resultLabel;
@property (weak, nonatomic) IBOutlet UIButton *backToBankBtn;
@property (weak, nonatomic) IBOutlet UITextView *responseTextView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *btnGoShoppingTrailingLC;
@property (strong, nonatomic) TermsViewController *termsViewController;
@property (strong, nonatomic) NSDictionary *resultDict;
@property (assign, nonatomic) BindStatus bindStatus;

@end

@implementation TokenPushViewController

#pragma mark - Life Cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (_token.length > 0) {
        if (_termsViewController == nil) {
            dispatch_async(dispatch_get_main_queue(), ^{
                self.termsViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"TermsViewController"];
                self.termsViewController.delegate = self;
                self.termsViewController.modalPresentationStyle = UIModalPresentationFullScreen;
                [self presentViewController:self.termsViewController animated:false completion:nil];
            });
        }
    }else {
        [self configResultUIWithStatus:BindStatusCancel];
    }
}

#pragma mark - Private Method

- (void)configResultUIWithStatus:(BindStatus)status {
    switch (status) {
        case BindStatusSuccess:
        {
            [_statusImageView setImage:[UIImage imageNamed:@"success"]];
            [_resultLabel setText:@"Add Card Success"];
            [_responseTextView setHidden:false];
            [_responseTextView setText:[NSString stringWithFormat:@"%@", _resultDict]];
        }
            break;
        case BindStatusFail:
        {
            [_statusImageView setImage:[UIImage imageNamed:@"failed"]];
            [_resultLabel setText:@"Add Card Fail"];
            [_responseTextView setHidden:false];
            [_responseTextView setText:[NSString stringWithFormat:@"%@", _resultDict]];
        }
            break;
        case BindStatusCancel:
        {
            [_statusImageView setImage:[UIImage imageNamed:@"success"]];
            [_resultLabel setText:@"Cancel Success"];
            [_responseTextView setHidden:true];
            [_responseTextView setText:@""];
        }
            break;
        default:
            break;
    }
    
    if ([[self.resultDict objectForKey:@"callback_url"] length] <= 0 && _cancelUrl.length <= 0) {
        [_backToBankBtn setHidden:true];
        _btnGoShoppingTrailingLC.constant = (CGRectGetWidth(self.view.bounds) - CGRectGetWidth(_backToBankBtn.frame))/2;
    }else {
        [_backToBankBtn setHidden:false];
        _btnGoShoppingTrailingLC.constant = DEFAULT_TRAILING;
    }
}

- (void)doPushTokenize {
    [self setIndicatorHidden:false];
    [self pushTokenizeWithToken:_token successCallback:^(NSDictionary *result) {
        self.token = @"";
        self.bindStatus = BindStatusSuccess;
        self.resultDict = result;
        dispatch_async(dispatch_get_main_queue(), ^{
            [self setIndicatorHidden:true];
            [self configResultUIWithStatus:BindStatusSuccess];
        });
    } failureCallback:^(NSDictionary *result, NSError *error) {
        self.token = @"";
        self.bindStatus = BindStatusFail;
        if (result) {
            self.resultDict = result;
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [self setIndicatorHidden:true];
            [self configResultUIWithStatus:BindStatusFail];
        });
    }];
}

#pragma mark - IBAction

- (IBAction)backToBankPressed:(id)sender {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSURL *url = (self.bindStatus == BindStatusCancel) ? [NSURL URLWithString:self.cancelUrl] : [NSURL URLWithString:[self.resultDict objectForKey:@"callback_url"]];
        [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
    });
}

#pragma mark - Push Tokenize API

- (void)pushTokenizeWithToken:(NSString *)token
              successCallback:(void (^)(NSDictionary *result))successCallback
              failureCallback:(void (^)(NSDictionary *result,  NSError *error))failureCallback{
    
    NSURL *url = [NSURL URLWithString:PUSH_TOKENIZE_PATH];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    [request setHTTPMethod:@"POST"];
    [request setTimeoutInterval:10.0];
    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request addValue:PARTNER_KEY forHTTPHeaderField:@"x-api-key"];
    [request setCachePolicy:NSURLRequestReloadIgnoringLocalCacheData];
    
    NSDictionary *parametersDict = @{@"partner_key":PARTNER_KEY,
                                     @"tsp_push_token":token};
    NSData *parameterData =[NSJSONSerialization dataWithJSONObject:parametersDict options:NSJSONWritingPrettyPrinted error:NULL];
    [request setHTTPBody:parameterData];
    
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    
    [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {

        NSDictionary *JsonObject = (data != nil) ?[NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil] : nil;
        NSLog(@"%@", JsonObject);
        
        if (error) {
            if (failureCallback) {
                failureCallback(nil, error);
            }
        }else {
            if ([[JsonObject objectForKey:@"status"] integerValue] == 0 || [[JsonObject objectForKey:@"status"] integerValue] == 19004) {
                if (successCallback) {
                    successCallback(JsonObject);
                }
            }else {
                if (failureCallback) {
                    failureCallback(JsonObject, nil);
                }
            }
        }
    }] resume];
}

#pragma mark - TermsViewController Delegate

- (void)didFinishReadingTerms:(TermsViewController *)controller {
    [self doPushTokenize];
}

- (void)didCancelReadingTerms:(TermsViewController *)controller {
    _bindStatus = BindStatusCancel;
    [self configResultUIWithStatus:BindStatusCancel];
}

@end

File: Swift/TapPayTokenPushExample/TapPayTokenPushExample/TokenPushViewController.swift
//
//  TokenPushViewController.swift
//  TapPayTokenPushExample
//
//  Created by Cherri_TapPay_LukeCho on 2022/8/12.
//

import Foundation
import UIKit

private let DEFAULT_TRAILING = 59.0
private let PUSH_TOKENIZE_PATH = "https://sandbox.tappaysdk.com/tpc/tsp/token/push-tokenize"
private let PUSH_TOKENIZE_ACCESS_KEY = "axgLG8z7vVa8mgIzdU1233rjBFAEmTti19NZd1pI"

enum BindStatus {
    case success
    case fail
    case cancel
}

class TokenPushViewController : BaseViewController{
    
    @IBOutlet weak var statusImageView: UIImageView!
    @IBOutlet weak var resultLabel: UILabel!
    @IBOutlet weak var btnGoShoppingTrailingLC: NSLayoutConstraint!
    @IBOutlet weak var backToBankBtn: UIButton!
    @IBOutlet weak var responseTextView: UITextView!
    private var bindStatus: BindStatus?
    private var resultDict : Dictionary<String, Any>?
    var pushToken : String = ""
    var cancelUrl : String = ""
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if pushToken.count > 0 {
            let termsViewController = self.storyboard?.instantiateViewController(withIdentifier: "TermsViewController") as! TermsViewController
            termsViewController.delegate = self
            termsViewController.modalPresentationStyle = UIModalPresentationStyle.fullScreen
            DispatchQueue.main.async {
                self.present(termsViewController, animated: false)
            }
        }else {
            configResultUIWithStatus(status: .cancel)
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }
    
    private func configResultUIWithStatus(status: BindStatus) {
        switch status {
        case .success:
            statusImageView.image = UIImage.init(named: "success")
            resultLabel.text = "Add Card Success"
            responseTextView.isHidden = false
            responseTextView.text = "\(resultDict ?? Dictionary.init())"
        case .fail:
            statusImageView.image = UIImage.init(named: "failed")
            resultLabel.text = "Add Card Fail"
            responseTextView.isHidden = true
            responseTextView.text = "\(resultDict ?? Dictionary.init())"
        case .cancel:
            statusImageView.image = UIImage.init(named: "success")
            resultLabel.text = "Cancel Success"
            responseTextView.isHidden = true
            responseTextView.text = ""
        default:
            break
        }
        
        if (resultDict != nil) {
            let callbackUrl = resultDict!["callback_url"] as? String ?? ""
            if callbackUrl.count <= 0 && cancelUrl.count <= 0 {
                backToBankBtn.isHidden = true
                btnGoShoppingTrailingLC.constant = (self.view.bounds.width - backToBankBtn.frame.width)/2
            }else {
                backToBankBtn.isHidden = false
                btnGoShoppingTrailingLC.constant = DEFAULT_TRAILING
            }
        }
    }
    
    @IBAction func backToBankPressed(_ sender: Any) {
        DispatchQueue.main.async {
            let url = (self.bindStatus == .cancel) ? URL.init(string: self.cancelUrl)! : URL.init(string: self.resultDict!["callback_url"] as! String)!
            UIApplication.shared.open(url)
        }
    }
    
    private func pushTokenizeWithToken(token: String ,success: @escaping (_ result: Dictionary<String, Any>) -> Void ,fail: @escaping (_ result:Dictionary<String, Any>?, _ error:Error?) -> Void) {
        let identifier =  UIApplication.shared.beginBackgroundTask(withName: "Push_Token") as UIBackgroundTaskIdentifier
        let url = URL.init(string: PUSH_TOKENIZE_PATH)!
        var request = URLRequest.init(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(PUSH_TOKENIZE_ACCESS_KEY, forHTTPHeaderField: "x-api-key")
        
        let encoder = JSONEncoder()
        let parametersDict = ["partner_key":PUSH_TOKENIZE_ACCESS_KEY,"tsp_push_token":token]
        let data = try? encoder.encode(parametersDict)
        request.httpBody = data
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            UIApplication.shared.endBackgroundTask(identifier)
            if let data = data {
                do {
                    let json = try JSONSerialization.jsonObject(with: data)
                    let dic = json as! Dictionary<String,Any>
                    print(dic)
                    
                    let status = dic["status"] as! Int
                    if status == 0 || status == 19004 {
                        success(dic)
                    }else {
                        fail(dic, nil)
                    }
                } catch  {
                    print(error)
                    fail(nil, error)
                }
            }
        }.resume()
    }
    
    private func doPushTokenize() {
        setIndicatorHidden(hidden: false)
        pushTokenizeWithToken(token: pushToken) { result in
            self.pushToken = ""
            self.bindStatus = .success
            self.resultDict = result
            DispatchQueue.main.async {
                self.setIndicatorHidden(hidden: true)
                self.configResultUIWithStatus(status: .success)
            }
        } fail: {result,error in
            self.pushToken = ""
            self.bindStatus = .fail
            self.resultDict = result
            DispatchQueue.main.async {
                self.setIndicatorHidden(hidden: true)
                self.configResultUIWithStatus(status: .fail)
            }
        }
    }
}

extension TokenPushViewController: TermsViewControllerDelegate {
    
    func didFinishReadingTerms(controller: TermsViewController) {
        doPushTokenize()
    }
    
    func didCancelReadingTerms(controller: TermsViewController) {
        self.pushToken = ""
        self.bindStatus = .cancel
        configResultUIWithStatus(status: .cancel)
    }
}
